sympy/matrices/expressions/hadamard.py

```diff
@@ -119,11 +130,127 @@ def validate(*args):
         if A.shape != B.shape:
             raise ShapeError("Matrices %s and %s are not aligned" % (A, B))
 
-rules = (unpack,
-         flatten)
 
-canonicalize = exhaust(condition(lambda x: isinstance(x, HadamardProduct),
-                                 do_one(*rules)))
+def canonicalize(x):
+    """Canonicalize the Hadamard product ``x`` with mathematical properties.
+
+    Examples
+    ========
+
+    >>> from sympy.matrices.expressions import MatrixSymbol, HadamardProduct
+    >>> from sympy.matrices.expressions import OneMatrix, ZeroMatrix
+    >>> from sympy.matrices.expressions.hadamard import canonicalize
+
+    >>> A = MatrixSymbol('A', 2, 2)
+    >>> B = MatrixSymbol('B', 2, 2)
+    >>> C = MatrixSymbol('C', 2, 2)
+    >>> Z = ZeroMatrix(2, 2)
+    >>> O = OneMatrix(2, 2)
+
+    Hadamard product associativity:
+
+    >>> X = HadamardProduct(A, HadamardProduct(B, C))
+    >>> canonicalize(X)
+    A.*B.*C
+
+    Hadamard product commutativity:
+
+    >>> X = HadamardProduct(A, B)
+    >>> Y = HadamardProduct(B, A)
+    >>> canonicalize(X) == canonicalize(Y)
+    True
```

Comment:
I think it would be clearer to show what is actually returned by canonicalize(X). It can be done on one line though:
```
>>> canonicalize(HadamardProduct(A, B))
...
>>> canonicalize(HadamardProduct(B, A))
...